federatedscope/cl/model/SimCLR.py

```diff
@@ -0,0 +1,225 @@
+        out = self.layer1(out)
+        out = self.layer2(out)
+        out = self.layer3(out)
+        out = F.adaptive_avg_pool2d(out, (1, 1))
+        out = out.view(out.size(0), -1)
+        if(self.train_sup):
+            out = self.linear(out)
+        return out
+
+
+def get_block(block):
+    if(block=="BasicBlock"):
+        return BasicBlock
+    elif(block=="Bottleneck"):
+        return Bottleneck
+
+def ResNet18(num_classes=10, block="BasicBlock"):
+    return ResNet(get_block(block), [2,2,2,2], num_classes=num_classes)
+
+def ResNet34(num_classes=10, block="BasicBlock"):
+    return ResNet(get_block(block), [3,4,6,3], num_classes=num_classes)
+
+def ResNet56(num_classes=10, block="BasicBlock"):
+    return ResNet_basic(get_block(block), [9,9,9], num_classes=num_classes)
+
+
+### Retrieval function for backbones ###
+def create_backbone(name, num_classes=10, block='BasicBlock'):
+    if(name == 'VGG'):
+        net = VGGmodel(num_classes=num_classes)
+    elif(name == 'res18'):
+        net = ResNet18(num_classes=num_classes, block=block)
+    elif(name == 'res34'):
+        net = ResNet34(num_classes=num_classes, block=block)
+    elif(name == 'res56'):
+        net = ResNet56(num_classes=num_classes, block=block)
+
+    return net
+
+
+# SimCLR model 
+
+
+# Projector 
+class projection_MLP_simclr(nn.Module):
+    def __init__(self, in_dim, hidden_dim=512, out_dim=512):
+        super(projection_MLP_simclr, self).__init__()
+        self.layer1 = nn.Linear(in_dim, hidden_dim, bias=False)
+        self.layer1_bn = nn.BatchNorm1d(hidden_dim, affine=True)
+        self.layer2 = nn.Linear(hidden_dim, out_dim)
+        self.layer2_bn = nn.BatchNorm1d(out_dim, affine=False)
+
+    def forward(self, x):
+        x = F.relu(self.layer1_bn(self.layer1(x))) 
+        x = self.layer2_bn(self.layer2(x)) 
+        return x
+
+# SimCLR
+class simclr(nn.Module):
+    def __init__(self, bbone_arch):
+        super(simclr, self).__init__()
+        self.T = 0.5
+        self.register_buffer("rounds_done", torch.zeros(1))
+
+        self.backbone = create_backbone(bbone_arch, num_classes=0)
+        self.projector = projection_MLP_simclr(self.backbone.output_dim, hidden_dim=512, out_dim=512)
+
+    def forward(self, x1, x2, x3=None, deg_labels=None):
+        N = x1.shape[0]
+        z1, z2 = self.projector(self.backbone(x1)), self.projector(self.backbone(x2))
+#         L = NT_xentloss(z1, z2, temperature=self.T)
+
+        return z1, z2
+    
+def ModelBuilder(model_config, local_data):
+     # You can also build models without local_data
+    data = next(iter(local_data['train']))
+    if model_config.type == "SimCLR":
+        model = simclr(bbone_arch='res18')
+        return model
+    if model_config.type == "SimCLR_linear":
+        model = create_backbone(name='res18', num_classes=10)
+        pretrained_model = torch.load('checkpoint/SimCLR_on_Cifar4CL_lr0.5_lstep5_rn100.ckpt', map_location='cpu')
```

Comment:
it is not allowed to hv such a hard-coded path

Reply:
it updates and merge in general type